### DISCUSSION TITLE
Can I build an index from non-HTML sources?

### DISCUSSION THREAD

**Discussion by documentat-alibaev** [Pagefind/pagefind#582] [General]
I want to index my content without using `data-pagefind-*` options. In fact, I would like to index Markdown sources before the HTML site is even built. That way, I would be able to hide or even add some content to the index based on my own business logic.

Is there a way to do so? I cannot find documentation for the index file format or any CLI for generating it from non-HTML sources.

**Comment by documentat-alibaev** [comment]
Oh, I think I found it: https://pagefind.app/docs/node-api/

I was expecting a CLI, but there is only a JS API, is that correct?

**Reply by bglw** [reply]
Yes, the Node API is the intended path for pulling in non-HTML sources :) 

It uses the Pagefind binary under the hood in a special "service" mode, so there's no particular technical barrier to providing a CLI interface, just a design/implementation body of work. But the Node API will likely always be a superset of the functionality available, so the initial development work went there.

Let me know how it works out for you! The Node API / service mode is one of the newer areas of Pagefind, so I'm still looking out for general feedback on its direction 😄

**Comment by documentat-alibaev** [comment]
@bglw When using `addCustomRecord()`, is there a way to let Pagefind know about H2+ headers on the page, so that it would produce results with anchored links?

**Reply by bglw** [reply]
Not officially, no. Currently the expectation is that anything that is HTML will go through `addHTMLFile` for processing, and `addCustomRecord` is intended for _non_ HTML content that wouldn't have anchors.

That doesn't have to be a hard limitation, though! And there are other things that would be useful in the same vein, primarily setting weights for sections of content in the custom record.

There would be a way to have it work right now, but I'm hesitant to broadcast it since it would rely on some of Pagefind's internal tokens that aren't part of a stable API, so it would break in future releases. The main barrier to adding this as an official feature is figuring out the syntax for `addCustomRecord` to set custom weights and anchors.

**Reply by bglw** [reply]
A first cut might look like adding a `content_parts` option as an alternative to `content`. How does the dx of the following feel to you?

```js
await index.addCustomRecord({
    url: "/contact/",
    content_parts: [
        {
            content: "Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.",
            weight: 1
        },
        {
            content: "Tristique Adipiscing Nullam Vulputate Venenatis.",
            weight: 12,
            anchor: "tristique-adipiscing"
        },
        {
            content: "Maecenas sed diam eget risus varius blandit sit amet non magna.",
            weight: 1
        }
    ],
    language: "en",
    meta: {
        title: "Contact",
        category: "Landing Page"
    },
    filters: {
        tags: ["landing", "company"]
    },
    sort: {
        weight: "20"
    }
});
```

**Reply by documentat-alibaev** [reply]
I haven't touched weights, filters and sorting yet, so i don't quite know everything that is shown here. But as for `content_parts` — yes, this looks good and similar enough to the internal structure that I hope to feed to Pagefind. Except your snippet does not include explicit titles for the parts, which I think would be necessary.

**Reply by bglw** [reply]
I just copied that snippet from the docs — you can safely ignore the meta/filters/sort keys as they're all optional :)

> Except your snippet does not include explicit titles for the parts

Internally, Pagefind is non-hierarchical, so the example above would correspond to the HTML:

```html
<p>Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum</p>
<h2 id="tristique-adipiscing">Tristique Adipiscing Nullam Vulputate Venenatis</h2>
<p>Maecenas sed diam eget risus varius blandit sit amet non magna</p>
```

That is, the content for a part with an `anchor` is inherently the title for that anchor, and a non-titled anchor would be added without content:
```js
{ anchor: "tristique-adipiscing" }
```

The main thing is that anchors don't wrap content in Pagefind, they exist as markers in the flat content.

**Reply by documentat-alibaev** [reply]
Ok, got it. Looking forward to `content_parts`!

**Reply by documentat-alibaev** [reply]
Until `content_parts` are implemented, the best way to do what I want would be to wrap text into fake pages in runtime and pass them to `addHTMLFile()`, right?

**Reply by bglw** [reply]
Yeah, that would be the least fragile way to pipe things through as-is
